### PR TITLE
Add CI for pull requests and when code lands on main

### PR DESCRIPTION
**Pull request by lance**
When code lands on main, the images are published with the `tip` tag.
When a pull request is submitted smoke tests are run with the func client
API. A function is created for each of the builtin templates for the
buildpack runtimes (http/events for go/typescript), and then built.

Needs: knative-sandbox/kn-plugin-func#495

Signed-off-by: Lance Ball <lball@redhat.com>

### CODE REVIEW

File: .github/workflows/pull-requests.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+on:
+  [pull_request]
+
+jobs:
+  make-test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+        with:
+          fetch-depth: 0
```
Comment (lkingland) [inline .github/workflows/pull-requests.yaml:12]: This is a good idea; it might speed up our other CI as well.  Added it to the list from our related thread.

File: .github/workflows/pull-requests.yaml [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+      - name: Install pack
+        run: |
+          curl -s -L -o pack.tgz ${{ steps.pack-download-url.outputs.result }}
+          tar -xvf pack.tgz
+          
+      - name: Build
+        run: PACK_CMD="$(pwd)/pack" make
+
+      - name: Install func
```
Comment (lkingland) [inline .github/workflows/pull-requests.yaml:43]: This may not be necessary since we're not really concerned with the CLI in this package.

File: test/test_buildpacks.go [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+	},
+}
+
+func runTests(ctx context.Context, version string) error {
+
+	os.Setenv("FUNC_REGISTRY", repo)
+
+	for i := range buildpacks {
+		buildpacks[i].Buildpacks[0] = fmt.Sprintf("%s:%s", buildpacks[i].Buildpacks[0], version)
+	}
```
Comment (lkingland) [inline test/test_buildpacks.go:60]: This could be replaced with a function which returns a fully populated slice of test cases, removing the need for a global variable and the post-initialization version-append.  Something perhaps like
```
type testCase struct {
    Buildpacks []string
    Runtimes   []string
}
func testCases(version string) []testCase {
    ...
}
```

File: test/test_buildpacks.go [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+	// just a counter to avoid name collision
+	var fnNo int
+
+	for _, funcBinary := range funcBinaries {
+		for _, buildpack := range buildpacks {
+			for _, runtime := range buildpack.Runtimes {
+				for _, template := range templates {
+					fnName := fmt.Sprintf("fn-%s-%s-%d", runtime, template, fnNo)
+					if !runTest(ctx, packCmd, funcBinary, runtime, template, fnName, buildpack.Buildpacks) {
```
Comment (lkingland) [inline test/test_buildpacks.go:105]: This is _so_ close to a table-driven test!  It might be worth the effort to go all the way through, unless I am missing an important reason for using the binaries.  Perhaps I will try to write up an exploration of that possibility.

**Review by lkingland — APPROVED** [review]
This looks good to me!

I do have one suggestion with regard to the tests and how they are run.  This may be more than we want to tackle at this point, or there may be a requirement I'm not seeing that preclude.  However, below is the suggestion in note form, should they be useful down the road.

tl;dr  Utilize Go's inbuilt testing mechanisms along with the Functions client library.


We may be able to streamline the tests a bit by considering two changes:  Using Go's built in testing facilities (which offer, among other things, parallelization, panic recovery, configurable timeouts, per-test logging, race detection, standardized output logging, etc) and directly utilizing the Functions client library to avoid using binaries when not necessary (both simpler and removes a dependency on kn-plugin-func when what we are really after is to know that a Functions Client instance can create a Function using these packs.)

The above is a bit general and hand-wavy.  So at the risk of the pendulum swinging too far the other direction, it could look something like:
1.  Removing `Install Func` stanza from `pull-requests.yaml`
2. Replacing `go run...` with `go test ./...` in the Makefile's `test` target
3. Renaming `test_buildpacks.go` to `packs_test.go`
4. Refactoring `packs_test.go` to be `package packs` with a `TestPacks(t *testing.T)` in place of `main`
   (this should remove the need for the contexts, cancels, signal listeners in the goroutine (L21 - L30)
6. Replacing `runTests` and `runTest` with a table-driven test that utilizes Go's `t.Run`.  
   [Here's an example](https://eleni.blog/2019/05/11/parallel-test-execution-in-go/) I found after a quick search which shows what seems (at first glance at least) to be a decent example of using table-driven testing along with `t.Run`.   This has quite a few advantages, In addition to enabling parallelization it should be able to reduce the lines required here because, by leveraging the Go test framework, it already includes printing success and failure status, start/stop statements, test duration,  sampling for OS signals such as ^C and context cancellation.  It also goes further in recovering from panics, allowing for configurable timeouts, race detection, verbose/quiet logging and things like that.
7.  Updating all `fmt.*` to use `t.Log`
8. Replace the call to `funcBinary` with an instance of `fn.Client`, which would be instantiated using the `WithRegistry("gcr.io/boson-project")` option and invoked from within a temp directory, perhaps like the tests do [in the client library itself](https://github.com/knative-sandbox/kn-plugin-func/blob/main/client_test.go#L47).

File: .github/workflows/pull-requests.yaml (hunk 2 above)
Comment (lkingland) [reply .github/workflows/pull-requests.yaml:43]: Update:  Added a wall of text regarding this in the final comment 😊

File: hack/install-func-snapshot.sh [hunk 5]
```diff
@@ -0,0 +1,22 @@
...
+
+# install latest from main branch
+TMP_DIR="$(mktemp -d)"
+cd "$TMP_DIR"
+echo "TODO: CHANGE THIS TO knative-sandbox WHEN https://github.com/knative-sandbox/kn-plugin-func/pull/485/files lands"
+git clone https://github.com/lance/func
+cd func
+git checkout lance/builtin-language-packs
+echo "Downloading pkger"
+go get github.com/markbates/pkger/cmd/pkger
```
[the remark this answers is not in the document]
Comment (lance) [reply hack/install-func-snapshot.sh:18]: Yes - I *think* I saw this behavior and that's why it's there. TBH, debugging github actions can be a pain sometimes with a bit of stabbing around at things, so it's possible this ended up there unnecessarily.

**Comment by lance** [comment]
@lkingland thanks for the thorough review, and especially for your suggestions re: testing. I'll take a stab at it.
